src/mpi/pt2pt/ibsend.c

```diff
@@ -92,22 +92,20 @@ int MPIR_Ibsend_impl(const void *buf, int count, MPI_Datatype datatype, int dest
     MPIR_Request *request_ptr, *new_request_ptr;
     ibsend_req_info *ibinfo = 0;
 
-    mpi_errno = MPIR_Bsend_isend(buf, count, datatype, dest, tag, comm_ptr, IBSEND, &request_ptr);
-    if (mpi_errno != MPI_SUCCESS)
-        goto fn_fail;
-    MPII_SENDQ_REMEMBER(request_ptr, dest, tag, comm_ptr->context_id);
-
-    /* FIXME: use the memory management macros */
-    ibinfo = (ibsend_req_info *) MPL_malloc(sizeof(ibsend_req_info), MPL_MEM_BUFFER);
-    ibinfo->req = request_ptr;
-    ibinfo->cancelled = 0;
-    mpi_errno = MPIR_Grequest_start(MPIR_Ibsend_query, MPIR_Ibsend_free,
-                                    MPIR_Ibsend_cancel, ibinfo, &new_request_ptr);
+    MPIR_Request *req = MPIR_Request_create(MPIR_REQUEST_KIND__PROXY, 0);
+    MPIR_ERR_CHKANDJUMP1(req == NULL, mpi_errno, MPI_ERR_OTHER, "**nomem", "**nomem %s",
+                         "Ibsend request");
+
+    mpi_errno = MPIR_Bsend_isend(buf, count, datatype, dest, tag, comm_ptr,
+                                 IBSEND, &req->u.proxy.real_request);
     MPIR_ERR_CHECK(mpi_errno);
-    /* The request is immediately complete because the MPIR_Bsend_isend has
-     * already moved the data out of the user's buffer */
-    MPIR_Grequest_complete(new_request_ptr);
-    MPIR_OBJ_PUBLISH_HANDLE(*request, new_request_ptr->handle);
+
+    MPIR_Object_set_ref(req, 1);
+    (req)->cc_ptr = &(req)->cc;
+    /* Ibsend is local-complete */
+    MPIR_cc_set(req->cc_ptr, 0);
+    req->comm = NULL;
+    MPIR_OBJ_PUBLISH_HANDLE(*request, req->handle);
 
```

Comment:
`MPIR_Request_create_complete` does all this right? Just need to add a reference to the `MPIR_Bsend_isend` returned `real_request` to it.

Reply:
`MPIR_Request_create_complete` just return the same `lw_req` that is not meant to be user visible. Here we need a user visible request.

Reply:
`lw_req` is user visible.

Reply:
Say user call two `MPI_IBsend`, they will need two different request handles that they can wait or cancel on it. Can `lw_req` satisfy that?

Reply:
> Say user call two `MPI_IBsend`, they will need two different request handles that they can wait or cancel on it. Can `lw_req` satisfy that?

No, but the same problem exists for `MPI_Isend`. Strictly speaking it is not standard compliant.

Reply:
Right it looks like there is only one `lw_req` per process.

Reply:
I see. I guess we can do the same for `MPI_Isend` and make the "cancel" a NOOP.

Do you recommend to use the lightweigt request here?

Reply:
Yes, I think so. I think user expectations about canceling a send operations are low, and given the upcoming deprecation in the standard, it's even less of a concern IMO.

Reply:
OK,  I'll make the change.